src/gfx/pal_packing.cpp

```diff
@@ -0,0 +1,524 @@
+	    : _assigned{std::forward<Ts>(elems)...}, _protoPals{&protoPals} {}
+
+private:
+	template<typename Inner, template<typename> typename Constness>
+	class Iter {
+	public:
+		friend class AssignedProtos;
+		// For `iterator_traits`
+		using difference_type = typename std::iterator_traits<Inner>::difference_type;
+		using value_type = ProtoPalAttrs;
+		using pointer = Constness<value_type> *;
+		using reference = Constness<value_type> &;
+		using iterator_category = std::forward_iterator_tag;
+
+	private:
+		Constness<decltype(_assigned)> *_array = nullptr;
+		Inner _iter{};
+
+		Iter(decltype(_array) array, decltype(_iter) &&iter) : _array(array), _iter(iter) {}
+		Iter &skipEmpty() {
+			while (_iter != _array->end() && !_iter->has_value()) {
+				++_iter;
+			}
+			return *this;
+		}
+
+	public:
+		Iter() = default;
+
+		bool operator==(Iter const &other) const { return _iter == other._iter; }
+		bool operator!=(Iter const &other) const { return !(*this == other); }
+		Iter &operator++() {
+			++_iter;
+			skipEmpty();
+			return *this;
+		}
+		Iter operator++(int) {
+			Iter it = *this;
+			++(*this);
+			return it;
+		}
+		reference operator*() const {
+			assert((*_iter).has_value());
+			return **_iter;
+		}
+		pointer operator->() const {
+			return &(**this); // Invokes the operator above, not quite a no-op!
+		}
+
+		friend void swap(Iter &lhs, Iter &rhs) {
+			swap(lhs._array, rhs._array);
+			swap(lhs._iter, rhs._iter);
+		}
+	};
+public:
+	using iterator = Iter<decltype(_assigned)::iterator, std::remove_const_t>;
+	iterator begin() { return iterator{&_assigned, _assigned.begin()}.skipEmpty(); }
+	iterator end() { return iterator{&_assigned, _assigned.end()}; }
+	using const_iterator = Iter<decltype(_assigned)::const_iterator, std::add_const_t>;
+	const_iterator begin() const {
+		return const_iterator{&_assigned, _assigned.begin()}.skipEmpty();
+	}
+	const_iterator end() const { return const_iterator{&_assigned, _assigned.end()}; }
+
+	/**
+	 * Assigns a new ProtoPalAttrs in a free slot, assuming there is one
+	 * Args are passed to the `ProtoPalAttrs`'s constructor
+	 */
+	template<typename... Ts>
+	void assign(Ts &&...args) {
+		auto freeSlot = std::find_if_not(
+		    _assigned.begin(), _assigned.end(),
+		    [](std::optional<ProtoPalAttrs> const &slot) { return slot.has_value(); });
+
+		if (freeSlot == _assigned.end()) { // We are full, use a new slot
+			_assigned.emplace_back(std::forward<Ts>(args)...);
+		} else { // Reuse a free slot
+			freeSlot->emplace(std::forward<Ts>(args)...);
+		}
+	}
+	void remove(iterator const &iter) {
+		iter._iter->reset(); // This time, we want to access the `optional` itself
+	}
+	void clear() { _assigned.clear(); }
+
+	bool empty() const {
+		return std::find_if(
+		           _assigned.begin(), _assigned.end(),
+		           [](std::optional<ProtoPalAttrs> const &slot) { return slot.has_value(); })
+		       == _assigned.end();
```

Comment:
Can all closing delimiters go on their own line when clang-format wraps them, like braces do?
```c++
		return std::find_if(
		           _assigned.begin(), _assigned.end(),
		           [](std::optional<ProtoPalAttrs> const &slot) { return slot.has_value(); }
		       ) == _assigned.end();
```

Reply:
Even if there's no `== _assigned.end()` after?

Reply:
Yeah, I wouldn't mind that style; but if you prefer to not spend vertical space like that, okay.

```c++
if (condition) {
	return std::some_call(
		with, a, lambda, arg, [this](int x) {
			return x == 42;
		}, and, more, args
	);
}
```

Reply:
No such option in Clang 13, so sadly we'll have to reject this. :T